### DISCUSSION TITLE
Has anyone gotten this working on a Mac recently?

### DISCUSSION THREAD

**Discussion by cwrichardson** [thyrlian/AndroidSDK#68] [Q&A] [answered]
I've followed along with [this Medium article](https://andresand.medium.com/android-emulator-on-docker-container-f20c49b129ef) from a few years ago, and it gets quite close but not all the way there.

I've written in some detail on [StackOverflow](https://stackoverflow.com/questions/72802842/run-thyrlian-androidsdk-docker-image-on-mac), but TL;DR, I seem to be able to successfully create new virtual devices, but they never display anything except a gray screen in VNC, and crash if you touch them.

Anyone using this on a Mac currently, and if so, what did you need to do to get it running?

Thanks!

**Comment by thyrlian** [comment, marked as answer]
Hi @cwrichardson 👋 

I use it on my Mac (with Intel CPU), and everything works.  It should not matter which host OS you run, the Docker engine is taking care of all virtualization work.

According to your description and logs, you were able to launch the container successfully, and even connected via a VNC client.  That's great, the image works.  But you had a problem running an emulator.

You tried with the same emulator that I've been tested: `system-images;android-25;google_apis;armeabi-v7a`, this emulator should work.

Did you launch the emulator with `-gpu swiftshader_indirect` option?
`emulator -avd armeabi-v7a -no-audio -no-boot-anim -accel on -gpu swiftshader_indirect &`

I've mentioned in the [README](https://github.com/thyrlian/AndroidSDK#how-to-run-emulator) that:

> Make sure that the required SDK packages are installed
> `sdkmanager "platform-tools" "platforms;android-<api_level>" "emulator"`

From your Stackoverflow question, it seems you have downloaded all the required packages.

I think you've done everything properly, I have no other advice, but maybe you can try:
* Download the latest AndroidSDK Docker image
* Don't mount any existing Android SDK directory, have a clean start, download all the required packages from scratch
* Wait long enough for the emulator to be booted

The error you've got is actually just a warning, nothing fatal, should not stop you from using it.
> `Warning: QXcbConnection: XCB error`

Please try again with the latest image and wait a bit longer.  Thank you for your patience.  Good luck!

**Reply by cwrichardson** [reply]
Thanks @thyrlian, and 👋 back at you :)

Actually, I was using `-no-audio -no-boot-anim -gpu off`. With `-accel on -gpu swiftshader_indirect` it does seem to (ultimately :) ) work, though it throws additional warnings:

```
INFO    | configAndStartRenderer: setting vsync to 60 hz
WARNING | cannot add library /opt/android-sdk/emulator/qemu/linux-x86_64/lib64/vulkan/libvulkan.so: failed
INFO    | added library /opt/android-sdk/emulator/lib64/vulkan/libvulkan.so
...
INFO    | Your emulator is out of date, please update by launching Android Studio:
 - Start Android Studio
 - Select menu "Tools > Android > SDK Manager"
 - Click "SDK Tools" tab
 - Check "Android Emulator" checkbox
 - Click "OK"
```

Now I have more questions :)

1) do I need to worry about these new info/warnings? And if so, how do I update Android Studio?
2) in both sets of warnings, there's this one:

`INFO    | Warning: XKeyboard extension not present on the X server ((null):0, (null))`

And, in point of fact, I can't use the keyboard :) Any idea how to fix that?

And finally, definitely a little off topic, but I thought you could help — now that it's up and running, how do I debug things in WebView? Is there a DevTools equivalent or something?

Thanks again for your help!

**Comment by thyrlian** [comment]
Happy to hear it finally works for you.

Concerning your questions:

> 1. do I need to worry about these new info/warnings? And if so, how do I update Android Studio?

No, you can ignore this type of warning.

> 2. in both sets of warnings, there's this one: `INFO | Warning: XKeyboard extension not present on the X server ((null):0, (null))`

In fact, this Docker image is designed for CI purposes, for any interactive operation, you should consider a faster emulator on your host machine.  Nevertheless, this warning seems a warning from Qt (which renders the emulator) to VNC.  You can find some help in this [thread](https://github.com/spyder-ide/spyder/issues/3713#issuecomment-261838179), e.g.: `export QT_XKB_CONFIG_ROOT=/usr/share/X11/xkb`.

Regarding how to debug in WebView, sorry, this is out of my knowledge.
